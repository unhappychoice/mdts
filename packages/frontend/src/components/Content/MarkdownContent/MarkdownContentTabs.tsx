import { Box, Tab, Tabs } from '@mui/material';
import React from 'react';
import { useSearchParams } from 'react-router-dom';

interface MarkdownContentTabsProps {
  viewMode: 'preview' | 'frontmatter' | 'raw';
  hasFrontmatter: boolean;
}

const MarkdownContentTabs: React.FC<MarkdownContentTabsProps> = ({ viewMode, hasFrontmatter }) => {
  const [searchParams, setSearchParams] = useSearchParams();

  const handleChange = (event: React.SyntheticEvent, newValue: 'preview' | 'frontmatter' | 'raw') => {
    searchParams.set('tab', newValue);
    setSearchParams(searchParams);
  };

  return (
    <Box
      sx={{
        paddingLeft: '24px',
        marginLeft: '-32px',
        marginRight: '-32px',
        borderBottom: 1,
        borderColor: 'divider',
      }}
    >
      <Tabs value={viewMode} onChange={handleChange} aria-label="view mode tabs">
        <Tab value="preview" label="Preview" />
        {hasFrontmatter &&
          <Tab value="frontmatter" label="Frontmatter" />
        }
        <Tab value="raw" label="Raw" />
      </Tabs>
    </Box>
  );
};

export default MarkdownContentTabs;
