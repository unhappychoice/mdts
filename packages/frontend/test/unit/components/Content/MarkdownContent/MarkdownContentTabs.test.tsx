import { fireEvent, render, screen } from '@testing-library/react';
import React from 'react';
import { BrowserRouter, useSearchParams } from 'react-router-dom';
import MarkdownContentTabs from '../../../../../src/components/Content/MarkdownContent/MarkdownContentTabs';

// Mock useSearchParams
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useSearchParams: jest.fn(),
}));

describe('MarkdownContentTabs', () => {
  const mockSetSearchParams = jest.fn();
  const mockSearchParams = new URLSearchParams();

  beforeEach(() => {
    (useSearchParams as jest.Mock).mockReturnValue([
      mockSearchParams,
      mockSetSearchParams,
    ]);
    mockSearchParams.delete('tab'); // Reset search params for each test
    mockSetSearchParams.mockClear();
  });

  test('renders preview and raw tabs by default', () => {
    const { asFragment } = render(
      <BrowserRouter>
        <MarkdownContentTabs viewMode="preview" hasFrontmatter={false} />
      </BrowserRouter>
    );
    expect(asFragment()).toMatchSnapshot();
    expect(screen.getByRole('tab', { name: /preview/i })).toBeInTheDocument();
    expect(screen.queryByRole('tab', { name: /frontmatter/i })).not.toBeInTheDocument();
    expect(screen.getByRole('tab', { name: /raw/i })).toBeInTheDocument();
  });

  test('renders frontmatter tab when hasFrontmatter is true', () => {
    render(
      <BrowserRouter>
        <MarkdownContentTabs viewMode="preview" hasFrontmatter={true} />
      </BrowserRouter>
    );
    expect(screen.getByRole('tab', { name: /preview/i })).toBeInTheDocument();
    expect(screen.getByRole('tab', { name: /frontmatter/i })).toBeInTheDocument();
    expect(screen.getByRole('tab', { name: /raw/i })).toBeInTheDocument();
  });

  test('sets search param to "frontmatter" when frontmatter tab is clicked', () => {
    render(
      <BrowserRouter>
        <MarkdownContentTabs viewMode="preview" hasFrontmatter={true} />
      </BrowserRouter>
    );
    fireEvent.click(screen.getByRole('tab', { name: /frontmatter/i }));
    expect(mockSetSearchParams).toHaveBeenCalledWith(expect.any(URLSearchParams));
    expect(mockSetSearchParams.mock.calls[0][0].get('tab')).toBe('frontmatter');
  });

  test('sets search param to "raw" when raw tab is clicked', () => {
    render(
      <BrowserRouter>
        <MarkdownContentTabs viewMode="preview" hasFrontmatter={false} />
      </BrowserRouter>
    );
    fireEvent.click(screen.getByRole('tab', { name: /raw/i }));
    expect(mockSetSearchParams).toHaveBeenCalledWith(expect.any(URLSearchParams));
    expect(mockSetSearchParams.mock.calls[0][0].get('tab')).toBe('raw');
  });

  test('sets search param to "preview" when preview tab is clicked', () => {
    render(
      <BrowserRouter>
        <MarkdownContentTabs viewMode="raw" hasFrontmatter={false} />
      </BrowserRouter>
    );
    fireEvent.click(screen.getByRole('tab', { name: /preview/i }));
    expect(mockSetSearchParams).toHaveBeenCalledWith(expect.any(URLSearchParams));
    expect(mockSetSearchParams.mock.calls[0][0].get('tab')).toBe('preview');
  });
});
