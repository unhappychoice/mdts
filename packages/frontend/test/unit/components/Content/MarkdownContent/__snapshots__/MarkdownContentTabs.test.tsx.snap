// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`MarkdownContentTabs renders preview and raw tabs by default 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-i3qfz4"
  >
    <div
      class="MuiTabs-root css-jhczcq-MuiTabs-root"
    >
      <div
        class="MuiTabs-scroller MuiTabs-fixed css-s2t35c-MuiTabs-scroller"
        style="overflow: hidden; margin-bottom: 0px;"
      >
        <div
          aria-label="view mode tabs"
          class="MuiTabs-list MuiTabs-flexContainer css-hzcega-MuiTabs-list"
          role="tablist"
        >
          <button
            aria-selected="true"
            class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary Mui-selected css-mobu7h-MuiButtonBase-root-MuiTab-root"
            role="tab"
            tabindex="0"
            type="button"
          >
            Preview
          </button>
          <button
            aria-selected="false"
            class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary css-mobu7h-MuiButtonBase-root-MuiTab-root"
            role="tab"
            tabindex="-1"
            type="button"
          >
            Raw
          </button>
        </div>
        <span
          class="MuiTabs-indicator css-1qltlow-MuiTabs-indicator"
          style="left: 0px; width: 0px;"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
